File: src/runtime/boxes/bb_span.c
/* _XSPNC    SPAN        longest prefix of chars in set (≥1) */
#include "bb_box.h"
#include <stdlib.h>
#include <string.h>


spec_t bb_span(void *zeta, int entry)
{
    span_t *ζ = zeta;
    spec_t SPAN;
    if (entry==α)                                                               goto SPAN_α;
    if (entry==β)                                                               goto SPAN_β;
    SPAN_α:         ζ->δ=0;                                                     
                    while (Δ+ζ->δ<Ω && strchr(ζ->chars,Σ[Δ+ζ->δ])) ζ->δ++;      
                    if (ζ->δ <= 0)                                              goto SPAN_ω;
                    SPAN = spec(Σ+Δ, ζ->δ); Δ += ζ->δ;                          goto SPAN_γ;
    SPAN_β:         Δ -= ζ->δ;                                                  goto SPAN_ω;
    SPAN_γ:                                                                     return SPAN;
    SPAN_ω:                                                                     return spec_empty;
}

span_t *bb_span_new(const char *chars)
{ span_t *ζ=calloc(1,sizeof(span_t)); ζ->chars=chars; return ζ; }

File: src/runtime/boxes/BbSpan.cs
// BbSpan.cs — SPAN: longest prefix of chars in set (must be ≥1)
// Mirrors src/runtime/boxes/bb_span.c
//
// α: scan forward while char in charset; if none matched → ω;
//    cursor += count; γ(count chars)
// β: cursor -= count; ω

namespace Snobol4.Runtime.Boxes;

public sealed class BbSpan : IByrdBox
{
    private readonly string _chars;
    private int _count;

    public BbSpan(string chars) { _chars = chars ?? ""; }

    public Spec Alpha(MatchState ms)
    {
        _count = 0;
        while (ms.CharInSet(ms.Cursor + _count, _chars)) _count++;
        if (_count <= 0) return Spec.Fail;
        var result = Spec.Of(ms.Cursor, _count);
        ms.Cursor += _count;
        return result;
    }

    public Spec Beta(MatchState ms)
    {
        ms.Cursor -= _count;
        return Spec.Fail;
    }
}
